Clang Address Sanitizer and Undefinded Behavior Sanitizer

mkuron opened the pull request:
Fixes #883

mkuron committed:
Add support for Clang sanitizers

mkuron committed:
Fixes per Clang ASAN and UBSAN

mkuron committed:
Enable ASAN and UBSAN in Clang CI build

mkuron committed:
Add tags to CI jobs

KaiSzuttor commented on CMakeLists.txt:
```diff
@@ -374,6 +381,18 @@ endif()
+  set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -g -O1")
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fsanitize=address -fno-omit-frame-pointer")
+endif()
+if( WITH_MSAN )
```
could you please reflect in the logic that ```WITH_ASAN``` and  ```WITH_MSAN``` are mutually exclusive?

mkuron replied:
Done

fweik commented on src/core/lbgpu.cpp:
```diff
@@ -258,7 +258,7 @@ void lb_reinit_parameters_gpu() {
     lbpar_gpu.mu[ii] = 0.0;
   
-    if (lbpar_gpu.viscosity[ii] > 0.0) {
+    if (lbpar_gpu.viscosity[ii] > 0.0 && lbpar_gpu.agrid > 0.0 && lbpar_gpu.agrid > 0.0 ) {
```
Duplicated condition

mkuron committed:
Some more ASAN/UBSAN fixes

mkuron committed:
Abort even when non-fatal undefined behavior is encountered

mkuron committed:
Sanitize some more undefined behavior

Mostly divisions by zero, but also some nullptrs bound to references and nonnull arguments

mkuron committed:
Merge branch 'python' of github.com:espressomd/espresso into sanitizer

mkuron committed:
Cleanup

KaiSzuttor commented:
if I compile with the MSAN option, I get
```ImportError: /tikhome/kai/espresso/build/src/core/io/writer/h5md/libH5mdCore.so: undefined symbol: __msan_retval_tls```

mkuron commented:
> `__msan_retval_tls`

I couldn't find any documentation on how to make MSAN work with shared libraries. It's possible that the Python interpreter needs to be recompiled with MSAN enabled.
It works fine though with the unit tests, so I'll leave the option in for now.

That means we are ready to merge now. Please review, @fweik @KaiSzuttor.

mkuron committed:
MSAN is experimental for now

fweik commented:
LGTM. @KaiSzuttor you can hit the button if you are happy.
